0016-consensus-parameters-change-proposal.md

```diff
@@ -0,0 +1,124 @@
+as possible without affecting the performance of the consensus layer.
+
+## Decision
+
+Implement governance proposal which changes consensus parameters only.
+
+## Implementation
+
+A new type of governance proposal named `ChangeParametersProposal` should be
+added to the consensus layer.
+
+```golang
+type ProposalContent struct {
+  ...
+  ChangeParameters *ChangeParametersProposal `json:"change_parameters,omitempty"`
+}
+```
+
+The proposal should contain a CBOR-encoded `Document` describing:
+
+- the name of the consensus application the changes apply to and
+
+- parameter changes which should be applied to the application's
+`ConsensusParameters` structure,
+
+and na `Epoch` at which the changes should be applied.
+
+```golang
+// ChangeParametersProposal is a consensus parameters change proposal.
+type ChangeParametersProposal struct {
+  // Document is a consensus parameters change document.
+  Document cbor.RawMessage `json:"document"`
+  // Epoch is the epoch at which the changes should be applied.
+  Epoch beacon.EpochTime `json:"epoch"`
+}
+```
+
+For example, a Tendermint consensus backend could define the document using
+the following structure.
+
+```golang
+// Document describes consensus parameter changes for an application.
+type Document struct {
+  cbor.Versioned
+
+  // App identifies the tendermint consensus application to which these changes
+  // should be apply to.
+  App uint8 `json:"app"`
+  // Changes are consensus parameter changes that should be applied to the
+  // application.
+  Changes cbor.RawMessage `json:"changes"`
+}
+```
+
+Once a `ChangeParametersProposal` is accepted the governance module will on
+specified epoch submit a message to the given application containing changes
+from the proposal. The module can notify the application using the existing
+message dispatch mechanism.
+
+Each application should carefully scope which parameters are allowed to be
+changed. For example, a Tendermint governance application could allow changing
+only the gas costs and the voting period.
+
+```golang
+// ConsensusParameterChanges define allowed governance consensus parameter
+// changes.
+type ConsensusParameterChanges struct {
+  cbor.Versioned
+
+  // GasCosts are the new gas costs.
+  GasCosts *transaction.Costs `json:"gas_costs,omitempty"`
+  // VotingPeriod is the new voting period.
+  VotingPeriod *beacon.EpochTime `json:"voting_period,omitempty"`
+}
+```
+
+How changes are executed is up to the application.
+
+```golang
+// ChangeParameters changes governance application consensus parameters.
+func (app *governanceApplication) ChangeParameters(changes *ConsensusParameterChanges) {
+  // Decode, validate, execute and apply changes.
+}
+```
+
+## Consequences
+
+### Positive
+
+- Agile and zero-downtime consensus parameter changes.
+
+- Separation of consensus parameter changes and consensus upgrades.
+
+### Negative
+
+- Consensus breaking change.
```

Comment:
The point below (not sure why that is listed under negative) would make this non-breaking (as it could be enabled later via governance, similar to PCS). I would suggest that you also include this in the ADR, e.g. adding a `EnableChangeParametersProposal` parameter to `ConsensusParameters` of the governance module and only supporting this new proposal type in case this is enabled (otherwise treating it as an invalid proposal kind).

Reply:
Negative as we have to add a new parameter. Too many flags makes code less readable and more complex.

Reply:
Ok, then explicitly mention the reason in the doc.